File: Data Structures & Algorithms/generate-parentheses/submission-2.js
class Solution {
    /**
     * @param {number} n
     * @return {string[]}
     */
    backtrack(openN, closedN, n, res, stack){
        if(openN === closedN && openN === n){
            res.push(stack);
            return;
        }

        if(openN < n){
            this.backtrack(openN + 1, closedN, n, res, stack + '(');
        }
        if(closedN < openN){
            this.backtrack(openN, closedN + 1, n, res, stack + ')');
        }
    }


    generateParenthesis(n) {
        const res = [];
        this.backtrack(0,0, n, res, '');
        return res;    
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-2.cpp
class Solution {
public:
    void zeroOutIsland(vector<vector<char>>& grid, int i, int j){
        if(i<0||j<0||i>=grid.size()||j>=grid[0].size()||grid[i][j]=='0'){
            return;
        }

        grid[i][j] = '0';
        zeroOutIsland(grid,i+1,j);
        zeroOutIsland(grid,i-1,j);
        zeroOutIsland(grid,i,j+1);
        zeroOutIsland(grid,i,j-1);
        return;
    }
    int numIslands(vector<vector<char>>& grid) {
        int res{};
        for(int i{}; i<grid.size();i++){
            for(int j{}; j<grid[0].size(); j++){
                if(grid[i][j]=='1'){
                    res++;
                    zeroOutIsland(grid,i,j);
                }
            }
        }
    
        return res;
    }




};


